docs/developer-docs/security/rust-canister-development-security-best-practices.md

```diff
@@ -461,6 +461,99 @@ Using the rust CDK, the reccuring timer is also lost on upgrade as explained in
+
+[More information.](https://internetcomputer.org/docs/current/developer-docs/integrations/https-outcalls/https-outcalls-how-it-works#compromised-replicas)
+
+[Data confidentiality security recommendations.](https://internetcomputer.org/docs/current/developer-docs/security/general-security-best-practices#data-confidentiality-on-the-internet-computer)
+
+### Ensure your canisters have a sufficiently large quota with the HTTP server
+
+#### Security concern
+
+When an HTTP outcall is performed it is amplified by the number of replicas in the subnet. The endpoint will receive not one request but as many requests as nodes in the subnet.
+
+Most web servers implement some sort of rate limiting, this is a mechanism used to restrict the number of requests a client can make to a web server within a specific time period, preventing abuse or excessive usage of their API(s).
+
+#### Recommendation
+
+You should consider such rate limits when designing and implementing your canisters. Rate limits are enforced using different time granularities, e.g., seconds or minutes. For second-granularity enforcement, make sure that the simultaneous requests by all subnet replicas do not violate the quota. Violations may lead to temporary or permanent bans.
+
+[More information.](https://internetcomputer.org/docs/current/developer-docs/integrations/https-outcalls/https-outcalls-how-it-works#rate-limiting-by-servers)
+
+### Only make HTTP outcall requests to idempotent endpoints
+
+#### Security concern
+
+As mentioned before, if an HTTP outcall is performed it is amplified by the number of replicas in the subnet. That means the queried endpoint will receive the same request several times. This is especially risky in requests that change the endpoint state given one HTTP outcall could lead to unintentionally changing the endpoint state several times.
+
+#### Recommendation
+
+Make sure the endpoints, called by an HTTP outcall, are idempotent, i.e. the queried endpoint has the same behaviour with the same request payload, no matter the number of times it is called.
+
+Some servers support the use of idempotency keys. These keys are random unique strings submitted in the HTTP request as headers. If used with the HTTP outcalls feature, all requests sent by each (honest) replica will contain the same idempotency key. This allows the server to recognize duplicated requests (i.e requests with the same idempotency key), handle just one and modify the server state only once. Note that this is a feature that must be supported by the server.
+
+[More information.](https://internetcomputer.org/docs/current/developer-docs/integrations/https-outcalls/https-outcalls-how-it-works#post-requests-must-be-idempotent)
+
+### Ensure HTTP responses are identical
+
+#### Security concern
+
+When replicas of a subnet receive HTTP responses, these responses must be identical. Otherwise the consensus won’t be achieved and the HTTP response will be rejected, but still charged.
```

Comment:
nit: 
```suggestion
When replicas of a subnet receive HTTP responses, these responses must be identical. Otherwise, consensus won’t be achieved and the HTTP response will be rejected, but still charged.
```